src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -1073,14 +1085,116 @@ private void PopulateParamsFromRestRequest(Dictionary<string, object?> parameter
+        {
+            IInputField schemaForItem = context.Selection.Field.Arguments[inputArgumentName];
+            Dictionary<string, HashSet<string>> fieldsToAuthorize = new();
+            object? parameters;
+            if (parametersDictionary.TryGetValue(inputArgumentName, out parameters))
             {
-                throw new DataApiBuilderException(
-                    message: "Unauthorized due to one or more fields in this mutation.",
-                    statusCode: HttpStatusCode.Forbidden,
-                    subStatusCode: DataApiBuilderException.SubStatusCodes.AuthorizationCheckFailed
-                );
+                PopulateMutationFieldsToAuthorize(fieldsToAuthorize, schemaForItem, entityName, context, parameters, _runtimeConfigProvider.GetConfig());
+            }
+
+            IEnumerable<string> entityNames = fieldsToAuthorize.Keys;
+
+            foreach(string entityNameInMutation in entityNames)
+            {
+                if (!_authorizationResolver.AreRoleAndOperationDefinedForEntity(entityNameInMutation, clientRole, operation))
+                {
+                    throw new DataApiBuilderException(
+                        message: $"The client has insufficient permissions to perform the operation on the entity: {entityNameInMutation}",
+                        statusCode: HttpStatusCode.Forbidden,
+                        subStatusCode: DataApiBuilderException.SubStatusCodes.AuthorizationCheckFailed
+                        );
+                }
+            }
+
+            foreach ((string entityNameInMutation, HashSet<string> fieldsInEntity) in fieldsToAuthorize)
+            {
+                if (!AreFieldsAuthorizedForEntity(clientRole, entityNameInMutation, EntityActionOperation.Create, fieldsInEntity))
+                {
+                    throw new DataApiBuilderException(
+                        message: $"The client has insufficient permissions on one or more fields in the entity: {entityNameInMutation} referenced in this mutation.",
+                        statusCode: HttpStatusCode.Forbidden,
+                        subStatusCode: DataApiBuilderException.SubStatusCodes.AuthorizationCheckFailed
+                    );
+                }
+            }
+            
+            return fieldsToAuthorize;
+        }
+
+        private void PopulateMutationFieldsToAuthorize(
+            Dictionary<string, HashSet<string>> fieldsToAuthorize,
+            IInputField schema,
+            string entityName,
+            IMiddlewareContext context,
+            object? parameters,
+            RuntimeConfig runtimeConfig)
+        {
+            InputObjectType schemaObject = ResolverMiddleware.InputObjectTypeFromIInputField(schema);
+            if (!fieldsToAuthorize.ContainsKey(entityName))
+            {
+                fieldsToAuthorize.Add(entityName, new HashSet<string>());
+            }
+
+            if (parameters is List<ObjectFieldNode> fields)
```

Comment:
can you elaborate on each of these conditions you are testing? e.g. when do you expect to processing `List<ObjectFieldNode>` versus an `ObjectFieldNode` or `List<IValueNode>`

Reply:
Added example and explaination